src/OpenLoco/Vehicles/Routing.cpp

```diff
@@ -0,0 +1,670 @@
+                auto* vehicle = entity->asVehicle();
+                if (vehicle == nullptr)
+                {
+                    continue;
+                }
+
+                if (vehicle->getFlags38() & (Vehicles::Flags38::unk_0 | Vehicles::Flags38::unk_2))
+                {
+                    continue;
+                }
+
+                if (vehicle->getTrackLoc() == interest.loc && vehicle->getVar2C().track == tad)
+                {
+                    addr<0x001135F88, uint16_t>() = 1;
+                    break;
+                }
+
+                if (vehicle->getTrackLoc() == nextLoc && vehicle->getVar2C().track == backwardTaD)
+                {
+                    addr<0x001135F88, uint16_t>() = 1;
+                    break;
+                }
+            }
+        }
+        return interest.trackAndDirection & (1 << 15);
+    }
+
+    // 0x004A2CE7
+    // Passes occupied state via 0x001135F88
+    static void setSignalsOccupiedState(const LocationOfInterestHashMap& hashMap)
+    {
+        for (const auto& interest : hashMap)
+        {
+            if (!(interest.trackAndDirection & (1 << 15)))
+            {
+                continue;
+            }
+
+            bool isOccupied = addr<0x001135F88, uint16_t>() & 1;
+            uint32_t flags = (1ULL << 31) | (isOccupied ? 8ULL : 9ULL);
+            sub_48963F(interest.loc, interest.tad(), interest.trackType, flags);
+        }
+    }
+
+    // 0x004A2D4C
+    static bool sub_4A2D4C(const LocationOfInterest& interest)
+    {
+        if (!(interest.trackAndDirection & (1 << 15)))
+        {
+            return false;
+        }
+
+        if (sub_48963F(interest.loc, interest.tad(), interest.trackType, (1ULL << 31) | (0xA)) & (1 << 0))
+        {
+            addr<0x001135F88, uint16_t>() |= (1 << 0);
+        }
+        else
+        {
+            addr<0x001135F88, uint16_t>() |= (1 << 1);
+        }
+        return true;
+    }
+
+    static void findAllUsableTrackInBlock(const LocationOfInterest& initialInterest, const FilterFunction filterFunction, LocationOfInterestHashMap& hashMap);
+
+    // 0x004A313B
+    // Iterates all individual tiles of a track piece to find tracks that need inspection
+    static void findAllUsableTrackPieces(const LocationOfInterest& interest, const FilterFunction filterFunction, LocationOfInterestHashMap& hashMap)
+    {
+        if (!(_1135FA6 & (1 << 2)))
+        {
+            findAllUsableTrackInBlock(interest, filterFunction, hashMap);
+            return;
+        }
+
+        const auto tad = interest.tad();
+        auto nextLoc = interest.loc;
+        if (tad.isReversed())
+        {
+            auto& trackSize = Map::TrackData::getUnkTrack(tad._data);
+            nextLoc += trackSize.pos;
+            if (trackSize.rotationEnd < 12)
+            {
+                nextLoc -= Map::Pos3{ _503C6C[trackSize.rotationEnd], 0 };
+            }
+        }
+
+        std::vector<LocationOfInterest> trackToCheck;
```

Comment:
Did the original code use the heap? May be a performance hit here.

Reply:
Yes it did use the heap but i don't think that is an issue